mplc/multi_partner_learning/fast_mpl.py

```diff
@@ -0,0 +1,568 @@
+
+    def generate_smodel(self):
+        smodel = tf.keras.models.Sequential()
+        smodel.add(tf.keras.Input(shape=(self.model.output_shape[-1],)))
+        smodel.add(NoiseAdaptationChannel(name='s-model'))
+        opt = self.model.optimizer.from_config(self.model.optimizer.get_config())
+        smodel.compile(optimizer=opt, loss='categorical_crossentropy', metrics=self.dataset.model_metrics_names[1:])
+        return smodel
+
+    def fit(self):
+        # TF function definition
+        @tf.function
+        def fit_pretrain_minibatch(model,
+                                   partners_minibatches,
+                                   partners_optimizers,
+                                   partners_weights,
+                                   aggregation_weights):
+            for p_id, minibatch in enumerate(partners_minibatches):  # minibatch == (x,y)
+                # minibatch[0] in a tensor of shape=(number of batch, batch size, img).
+                # We cannot iterate on tensors, so we convert this tensor to a list of
+                # *number of batch* tensors with shape=(batch size, img)
+                x_minibatch = tf.unstack(minibatch[0], axis=0)
+                y_minibatch = tf.unstack(minibatch[1], axis=0)  # same here, with labels
+
+                for model_w, partner_w in zip(model.trainable_weights,
+                                              partners_weights[p_id]):  # set the model weights to partner's ones
+                    model_w.assign(partner_w.read_value())
+
+                for x, y in zip(x_minibatch, y_minibatch):  # iterate over batches
+                    with tf.GradientTape() as tape:
+                        y_pred = model(x)
+                        loss = model.compiled_loss(y, y_pred)
+                    model.compiled_metrics.update_state(y, y_pred)  # log the loss and accuracy
+                    partners_optimizers[p_id].minimize(loss, model.trainable_weights,
+                                                       tape=tape)  # perform local optimization
+
+                for model_w, partner_w in zip(model.trainable_weights,
+                                              partners_weights[p_id]):  # update the partner's weights
+                    partner_w.assign(model_w.read_value())
+            # at the end of the minibatch, aggregate all the local weights
+            for i, weights_per_layer in enumerate(zip(*partners_weights)):
+                model.trainable_weights[i].assign(tf.tensordot(weights_per_layer, aggregation_weights, [0, 0]))
+
+            # inform the partners with the new weights
+            for p_id in range(len(partners_minibatches)):
+                for new_w, partner_w in zip(model.trainable_weights, partners_weights[p_id]):
+                    partner_w.assign(new_w.read_value())
+
+        @tf.function
+        def fit_minibatch(model, partners_minibatches, partners_optimizers, partners_weights, smodel_list,
```

Comment:
This looks very similar to the `fit_minibatch` function in FastFedAvg. Do you think it can be factorized?

Reply:
For the fit_minibatch function, as I said in previous review I need to declare it in the fit function (to redeclare it at each fit call)
Because of that I don’t see how I can define it anywhere else so the duplicate piece of code seems mandatory to me. Do you have any suggestion for improve that? It will be indeed much clearer